typed-racket-lib/typed-racket/typecheck/tc-lambda-unit.rkt

```diff
@@ -406,12 +406,45 @@
                        (tc/lambda-clause/check
                         f* b* dom (values->tc-results rng (formals->objects f*)) rst)]))])))))
 
-(define (tc/mono-lambda/type formals bodies expected)
-  (make-Fun
-    (tc/mono-lambda
-      (for/list ([f (in-syntax formals)] [b (in-syntax bodies)])
-         (list (make-formals f) b))
-      expected)))
+(define (tc/dep-lambda formalss-stx bodies-stx dep-fun-ty)
+  (match-define (DFun: dom rng) dep-fun-ty)
+  (define formalss (stx-map make-formals formalss-stx))
+  (define bodies (syntax->list bodies-stx))
+  (match* (formalss bodies)
+    [((list fs) (list body))
+     (cond
+       [(not (= (length (formals-positional fs))
+                (length dom)))
+        (tc-error/expr #:return dep-fun-ty
+                       (format "Expected ~a positional arguments, given ~a."
+                               (length dom)
+                               (length (formals-positional fs))))]
+       [(formals-rest fs)
+        (tc-error/expr #:return dep-fun-ty
+                       "Dependent functions do not currently support rest arguments.")]
+       [else
+        (define arg-names (formals-positional fs))
+        (define instantiated-deps
+          (for/list ([d (in-list dom)])
+            (instantiate-obj d arg-names)))
+        (with-extended-lexical-env
+            [#:identifiers arg-names
+             #:types instantiated-deps]
+          (tc-body/check body (values->tc-results rng (map -id-path arg-names))))
+        dep-fun-ty])]
+    [(fs bs)
+     (tc-error/expr #:return dep-fun-ty
```

Comment:
This code is unreachable at the moment right? Since there's no `case->d`.

Reply:
Yah -- it's more of an artifact of how "listy" all of our lambda/function related code is, I think (i.e. all of our functions are actually case->, etc)